cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java

```diff
@@ -711,29 +715,35 @@ private void addBranchPrograms(String node, WorkflowProgramAdder programAdder, b
   }
 
   private BatchPhaseSpec getPhaseSpec(String programName, PipelinePhase phase) {
+    Map<String, String> phaseConnectorDatasets = new HashMap<>();
     // if this phase uses connectors, add the local dataset for that connector if we haven't already
     for (StageSpec connectorInfo : phase.getStagesOfType(Constants.Connector.PLUGIN_TYPE)) {
       String connectorName = connectorInfo.getName();
-      String datasetName = connectorDatasets.get(connectorName);
-      if (datasetName == null) {
-        datasetName = "conn-" + connectorNum++;
+      if (!connectorDatasets.containsKey(connectorName)) {
+        String datasetName = "conn-" + connectorNum++;
         connectorDatasets.put(connectorName, datasetName);
+        phaseConnectorDatasets.put(connectorName, datasetName);
         // add the local dataset
         ConnectorSource connectorSource = new MultiConnectorSource(datasetName, null);
         connectorSource.configure(getConfigurer());
+      } else {
+        phaseConnectorDatasets.put(connectorName, connectorDatasets.get(connectorName));
       }
     }
+
     // create a local dataset to store alerts. At the end of the phase, the dataset will be scanned and alerts
     // published.
     for (StageSpec alertPublisherInfo : phase.getStagesOfType(AlertPublisher.PLUGIN_TYPE)) {
       String stageName = alertPublisherInfo.getName();
-      AlertPublisherSink alertPublisherSink = new AlertPublisherSink(stageName, null);
-      alertPublisherSink.configure(getConfigurer());
-    }
-
-    Map<String, String> phaseConnectorDatasets = new HashMap<>();
-    for (StageSpec connectorStage : phase.getStagesOfType(Constants.Connector.PLUGIN_TYPE)) {
-      phaseConnectorDatasets.put(connectorStage.getName(), connectorDatasets.get(connectorStage.getName()));
+      if (!connectorDatasets.containsKey(stageName)) {
```

Comment:
don't think this if-else is needed. stage names are guaranteed to be unique.

Reply:
But a stage might get used by multiple phases of programs, and this method is called when adding each phase of program, so we might already have the dataset name for the stage, this way we don't need to compute the dataset name again.

Reply:
ah ok got it.